kitsune/messages/utils.py

```diff
@@ -18,38 +18,42 @@ def send_message(to, text=None, sender=None):
     # We need a sender, a message, and someone to send it to
     if not sender or not text or not to:
         return
-    # This is users from the To field
-    users = {user for user in to if isinstance(user, User)}
-    # This is groups from the To field
-    groups = {group for group in to if isinstance(group, Group)}
-    # This is all the users that are going to receive the message
-    receivers = users | set(User.objects.filter(groups__in=groups).distinct())
+    # Prepare querysets for users and groups directly and efficiently
+    users_qs = User.objects.filter(
+        username__in=[obj.username for obj in to if isinstance(obj, User)]
+    )
+    groups_qs = Group.objects.filter(name__in=[obj.name for obj in to if isinstance(obj, Group)])
```

Comment:
This could be further optimized by populating a list of `usernames` and `group_names` in one go:

```python
username = []
group_names = []

for obj in to:
    if isinstance(obj, User):
        usernames.add(obj)
    else:
         group_names.add(obj)

group_qs = Group.objects.filter(name__in=group_names)
user_qs = User.objects.filter(username__in=usernames)
```
It would be even better when you construct the `to` iterable to make it something that you can manipulate easier. Eg a dictionary where you could have:
```python 
to = {
    'users': [list_of_usernames],
    'groups': [list_of_group_names]
```
this way you have essentially a flat list of strings instead of objects that's already separated - like using `values_list` instead of the whole object.

Reply:
any of these approaches will affect the following code too.